#Data Wrangling with Pandas

[Aj. NesT the Series](http://bit.ly/ajnesttheseriesSubscribe)

Referenence: [Pyhton Data Cleaning Cookbook 2020](https://github.com/PacktPublishing/Python-Data-Cleaning-Cookbook)

##Practice 2: การจัดการค่าที่หายไป การลบข้อมูลแถวที่ซ้ำกัน และการจัดระเบียบข้อมูลตามกลุ่ม

--> Managing Missing Values, Removing Duplicated Rows, and Organizing using Groupby 

[Python Pandas Cheat Sheets](https://www.enthought.com/wp-content/uploads/Enthought-Python-Pandas-Cheat-Sheets-1-8-v1.0.2.pdf)

##Workshop 6: ตรวจหาค่าที่หายไป ใส่ค่าข้อมูลใหม่ และลบแถวของข้อมูลที่หายไป
--> Finding, Inserting missing values and Deleting rows of missing values

###Input Datasets

* **[covidtotalswithmissings.csv](https://drive.google.com/file/d/1fZygbwQ5-8i-fFAMG-3BTur_hYf6S4V-/view?usp=sharing)** - The Covid-19 data were missing for Hong Kong 2020

####Question 6: เราจะตรวจหาข้อมูลที่หายไปแล้วทำการใส่ค่าเพื่อแก้ไขข้อมูลได้อย่างไร?

**STEP 1:** นำเข้า Datasets โดยทำการ Upload ไฟล์ covidtotalswithmissings.csv ขึ้น Google Colab (ไฟล์นี้จะถูกลบเมื่อเราหลุดจาก Session หรือปิด Browser) หรือ Python IDE เช่น PyCharm, VSCode, Juputer Notebook, etc.

**STEP 2:** นำเข้า pandas และ numpy libraries มาใช้งาน

**STEP 3:** ทำการอ่านไฟล์ .csv ด้วย pandas เก็บเป็น Dataframe

ใช้คำสั่ง .read_csv()

In [ ]:
#import pandas and numpy
import pandas as pd
import numpy as np
covidtotals = pd.read_csv("/content/covidtotalswithmissings.csv")
print(covidtotals)

    iso_code    lastdate        location  ...  median_age  gdp_per_capita  hosp_beds
0        AFG  2020-06-01     Afghanistan  ...        18.6        1803.987       0.50
1        ALB  2020-06-01         Albania  ...        38.0       11803.431       2.89
2        DZA  2020-06-01         Algeria  ...        29.1       13913.839       1.90
3        AND  2020-06-01         Andorra  ...         NaN             NaN        NaN
4        AGO  2020-06-01          Angola  ...        16.8        5819.495        NaN
..       ...         ...             ...  ...         ...             ...        ...
205      VNM  2020-06-01         Vietnam  ...        32.6        6171.884       2.60
206      ESH  2020-06-01  Western Sahara  ...        28.4             NaN        NaN
207      YEM  2020-06-01           Yemen  ...        20.3        1479.147       0.70
208      ZMB  2020-06-01          Zambia  ...        17.7        3689.251       2.00
209      ZWE  2020-06-01        Zimbabwe  ...        19.6        

**STEP 4:** ตั้งค่าและแสดง index และ size ของ covidtotalsmissings data

ใช้คำสั่ง set_index() ให้คอลัมน์ "iso_code" เป็นคอลัมน์ index

In [ ]:
#set up the cumulative and demographic columns กดแค่ครั้งเดียว 
covidtotals.set_index("iso_code", inplace=True) #ย้อนกลับไปให้ใช้ reset_index()
print(covidtotals)

            lastdate        location  ...  gdp_per_capita  hosp_beds
iso_code                              ...                           
AFG       2020-06-01     Afghanistan  ...        1803.987       0.50
ALB       2020-06-01         Albania  ...       11803.431       2.89
DZA       2020-06-01         Algeria  ...       13913.839       1.90
AND       2020-06-01         Andorra  ...             NaN        NaN
AGO       2020-06-01          Angola  ...        5819.495        NaN
...              ...             ...  ...             ...        ...
VNM       2020-06-01         Vietnam  ...        6171.884       2.60
ESH       2020-06-01  Western Sahara  ...             NaN        NaN
YEM       2020-06-01           Yemen  ...        1479.147       0.70
ZMB       2020-06-01          Zambia  ...        3689.251       2.00
ZWE       2020-06-01        Zimbabwe  ...        1899.775       1.70

[210 rows x 11 columns]


**STEP 5:** ตั้งค่าคอลัมน์ข้อมูลสะสม Cumulatice และคอลัมน์ข้อมูลประชากรศาสตร์ Demographic

In [ ]:
#set up the cumulative and demographic columns
totvars = ['location','total_cases','total_deaths','total_cases_pm',
  'total_deaths_pm']
demovars = ['population','pop_density','median_age','gdp_per_capita',
  'hosp_beds']
print(totvars)
print(demovars)

['location', 'total_cases', 'total_deaths', 'total_cases_pm', 'total_deaths_pm']
['population', 'pop_density', 'median_age', 'gdp_per_capita', 'hosp_beds']


**STEP 6:** เช็คค่า missing data ที่หายไปของคอลัมน์กลุ่ม demographic

Check the demographic columns for missing data.

(1) ใช้คำสั่ง .isnull().sum(axis=0) 

- 0 คือ เช็คหัว Columns ทำการนับจำนวน missing data 

- พบว่า population = 0 แสดงว่าไม่พบค่า missing data เป็น compltete data

- พบว่าข้อมูล hosp_beds = 46 ประเทศ พบค่า missing data คิดเป็นประมาณ 20% จากทั้งหมด 210 ประเทศ

(2) ใช้คำสั่ง .isnull().sum(axis=1) 
- 1 คือ เช็ค Rows ของ Index ทำการนับจำนวน missing data พบว่า AND มี missing data 3 ค่า

(3) ใช้คำสั่ง .value_counts() นับค่าจำนวน missing data ที่พบของ demovarmisscnt
- พบว่ามี 156 ประเทศ ข้อมูลครบสมบูรณ์ complete values

- พบว่า 24 ประเทศ มีข้อมูลหายไป 1 ค่า (missing data) 
- 12 ประเทศ มีข้อมูลหายไป 2 ค่า 10 ประเทศ มีข้อมูลหายไป 3 ค่า และ 8 ประเทศ มีข้อมูลหายไป 8 ค่า

In [ ]:
#check the demographic columns for missing
print(covidtotals[demovars].isnull().sum(axis=0))  #แสดง missing data ของ columns
demovarsmisscnt = covidtotals[demovars].isnull().sum(axis=1) #แสดง missing data ของ index rows
print() #ขึ้นบรรทัดใหม่
print(demovarsmisscnt)
print()
print(demovarsmisscnt.value_counts())

population         0
pop_density       12
median_age        24
gdp_per_capita    28
hosp_beds         46
dtype: int64

iso_code
AFG    0
ALB    0
DZA    0
AND    3
AGO    1
      ..
VNM    0
ESH    3
YEM    0
ZMB    0
ZWE    0
Length: 210, dtype: int64

0    156
1     24
2     12
3     10
4      8
dtype: int64


**STEP 7:** สร้างเงื่อนไขแสดงเฉพาะประเทศที่มีค่าที่หายไป หรือ missing data ตั้งแต่ 3 ค่าขึ้นไป 

List the countries with three or more missing values for the demographic data

ใช้คำสั่ง .loc[ rows, columns ] ทำการ Filter เลือกแถวและคอลัมน์ที่ต้องการ

In [ ]:
#check the demographic columns for missing --> condition >= 3
print(covidtotals.loc[demovarsmisscnt>=3, ['location'] + demovars].head(5)) #ลองใส่ 18 (10 + 8)

                                 location  ...  hosp_beds
iso_code                                   ...           
AND                               Andorra  ...        NaN
AIA                              Anguilla  ...        NaN
BES       Bonaire Sint Eustatius and Saba  ...        NaN
VGB                British Virgin Islands  ...        NaN
FRO                        Faeroe Islands  ...        NaN

[5 rows x 6 columns]


In [ ]:
print(covidtotals.loc[demovarsmisscnt>=3, ['location'] + demovars].head(5).T) #Transpost
type(demovarsmisscnt)

iso_code            AND       AIA  ...                     VGB             FRO
location        Andorra  Anguilla  ...  British Virgin Islands  Faeroe Islands
population        77265     15002  ...                   30237           48865
pop_density     163.755       NaN  ...                 207.973          35.308
median_age          NaN       NaN  ...                     NaN             NaN
gdp_per_capita      NaN       NaN  ...                     NaN             NaN
hosp_beds           NaN       NaN  ...                     NaN             NaN

[6 rows x 5 columns]


pandas.core.series.Series

**STEP 8:** ตรวจ Covid case data ของค่าที่หายไป

Check the Covid case data for missing values

(1) ใช้คำสั่ง .isnull().sum(axis=0) 

- 0 คือ เช็คหัว Columns ทำการนับจำนวน missing data 

(2) ใช้คำสั่ง .isnull().sum(axis=1) 
- 1 คือ เช็ค Rows ของ Index ทำการนับจำนวน missing data

(3) ใช้คำสั่ง .value_counts() นับค่าจำนวน missing data ที่พบของ totvarsmisscnt

- location, total_cases, totol_deaths = 0 คือไม่พบ missing data แสดงว่าข้อมูลครบ complete data

- พบว่ามี 2 columns คือ total_cases_pm, total_deaths_pm ที่มีค่า missing data คอลัมน์ละ 1 ค่า

In [ ]:
#check the cumulative columns for missing
print(covidtotals[totvars].isnull().sum(axis=0)) #แสดง columns จำนวน missing data
totvarsmisscnt = covidtotals[totvars].isnull().sum(axis=1) #แสดง index rows จำนวน missing data
print()
print(totvarsmisscnt)
print()
print(totvarsmisscnt.value_counts()) #นับจำนวนค่า missing data ที่พบของ totvarsmisscnt

location           0
total_cases        0
total_deaths       0
total_cases_pm     1
total_deaths_pm    1
dtype: int64

iso_code
AFG    0
ALB    0
DZA    0
AND    0
AGO    0
      ..
VNM    0
ESH    0
YEM    0
ZMB    0
ZWE    0
Length: 210, dtype: int64

0    209
2      1
dtype: int64


In [ ]:
#เงื่อนไข >0 นับจำนวน missing data ทั้งหมด เพื่อต้องการหาประเทศที่มี missing data (มีหายไป 2 ค่า 1 ประเทศ)
print(covidtotals.loc[totvarsmisscnt>0])  #ประเทศนั้นคือ HKG

            lastdate   location  ...  gdp_per_capita  hosp_beds
iso_code                         ...                           
HKG       2020-05-26  Hong Kong  ...        56054.92        NaN

[1 rows x 11 columns]


In [ ]:
#แสดงรายละเอียดให้มากขึ้นใช้ Transpost
print(covidtotals.loc[totvarsmisscnt>0].T) 
#สังเกตว่ามี total_cases_pm และ total_deaths_pm มีค่าเท่ากับ Nan

iso_code                 HKG
lastdate          2020-05-26
location           Hong Kong
total_cases                0
total_deaths               0
total_cases_pm           NaN
total_deaths_pm          NaN
population       7.49699e+06
pop_density          7039.71
median_age              44.8
gdp_per_capita       56054.9
hosp_beds                NaN


**STEP 9:** เปลี่ยนค่า NaN (Missing Values) ที่พบ ให้มีจำนวนเป็น 0 โดยต้องเติมค่าที่ถูกต้องใส่เข้าไปแทน

Use the fillna method to fix the missing cases data for the one country affected
(Hong Kong)

ใช้คำสั่ง .fillna() แทนที่ค่าที่หายไปด้วยค่าที่ผู้ใช้ระบุ

ตำแหน่ง missing data ที่ต้องการแทน.fillna(ค่าที่ต้องการแทน จะใส่ตัวเลข สูตรคำนวณ หรือ "ตัวอักษร" ก็ได้) 

In [ ]:
#use the fillna method to fix the mixing case data
#before use fillna()
print(covidtotals[totvars].isnull().sum(axis=0))
#.fillna ถ้ามีการแก้ไขต้องเริ่มรันใหม่ตั้งแต่ STEP 1 อีกครั้ง เช่น ลองใส่ .fillna(3) หรือ .fillna("Hello")
covidtotals.total_cases_pm.fillna(covidtotals.total_cases/(covidtotals.population/1000000), inplace=True)
covidtotals.total_deaths_pm.fillna(covidtotals.total_deaths/(covidtotals.population/1000000), inplace=True)
print()
#after use fillna()
print(covidtotals[totvars].isnull().sum(axis=0)) #แสดงค่าที่ใส่เข้าไปใหม่ 
covidtotals.to_csv("fill_covidtotals")

location           0
total_cases        0
total_deaths       0
total_cases_pm     1
total_deaths_pm    1
dtype: int64

location           0
total_cases        0
total_deaths       0
total_cases_pm     0
total_deaths_pm    0
dtype: int64


In [ ]:
#แสดงผลค่า values ของข้อมูล 'HKG' หลังจากใส่ค่าใหม่ที่แก้ไข missing data แล้ว
print(covidtotals.loc[['HKG']].T)

iso_code                 HKG
lastdate          2020-05-26
location           Hong Kong
total_cases                0
total_deaths               0
total_cases_pm             0
total_deaths_pm            0
population       7.49699e+06
pop_density          7039.71
median_age              44.8
gdp_per_capita       56054.9
hosp_beds                NaN


**STEP 10:** ลบแถวที่มีค่า NaN (Missing Values) ออก

Use the dropna method to remove the missing cases data

ใช้คำสั่ง .dropna() ลบแถวของค่าที่หายไป

In [ ]:
#Before use dropna
print(covidtotals)
print(covidtotals.shape)
#After use dropna
drop_covidtotals = covidtotals.dropna()
print(drop_covidtotals)
print(drop_covidtotals.shape)
drop_covidtotals.to_csv("drop_covidtotals")

            lastdate        location  ...  gdp_per_capita  hosp_beds
iso_code                              ...                           
AFG       2020-06-01     Afghanistan  ...        1803.987       0.50
ALB       2020-06-01         Albania  ...       11803.431       2.89
DZA       2020-06-01         Algeria  ...       13913.839       1.90
AND       2020-06-01         Andorra  ...             NaN        NaN
AGO       2020-06-01          Angola  ...        5819.495        NaN
...              ...             ...  ...             ...        ...
VNM       2020-06-01         Vietnam  ...        6171.884       2.60
ESH       2020-06-01  Western Sahara  ...             NaN        NaN
YEM       2020-06-01           Yemen  ...        1479.147       0.70
ZMB       2020-06-01          Zambia  ...        3689.251       2.00
ZWE       2020-06-01        Zimbabwe  ...        1899.775       1.70

[210 rows x 11 columns]
(210, 11)
            lastdate             location  ...  gdp_per_capita  hosp